### PR TITLE
ui: add React UI from upstream Prometheus

### PR DESCRIPTION
**Pull request by adrien-f**
* [X] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

* Add React UI accesible from the Query only for the moment as only the graph endpoint has been migrated.

I've been putting Thanos related modifications in the `ui/react-app/src/thanos` directory but there are some modifications in the root files. The goal is to be able to just copy paste the source from upstream and let the Thanos directory do the rest to ease upgrades.

## Verification

image

### CODE REVIEW

**Comment by bwplotka** [comment]
@adrien-f  can you rebase a bit?

@d-ulyanov @geobeau can you help in review?

File: pkg/ui/react-app/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+## Installing npm dependencies
+
+The React UI depends on a large number of [npm](https://www.npmjs.com/) packages. These are not checked in, so you will need to download and install them locally via the Yarn package manager:
+
+    yarn
+
+Yarn consults the `package.json` and `yarn.lock` files for dependencies to install. It creates a `node_modules` directory with all installed dependencies.
+
+## Running a local development server
```
Comment (geobeau) [inline pkg/ui/react-app/README.md]: This part should be adapted for Thanos

File: pkg/ui/react-app/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+## Building the app for production
+
+To build a production-optimized version of the React app to a `build` subdirectory, run:
+
+    yarn build
+
+**NOTE:** You will likely not need to do this directly. Instead, this is taken care of by the `build` target in the main Prometheus `Makefile` when building the full binary.
+
+## Integration into Prometheus
```
Comment (geobeau) [inline pkg/ui/react-app/README.md]: Prometheus -> Thanos

File: pkg/ui/react-app/public/manifest.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "Prometheus UI",
```
Comment (geobeau) [inline pkg/ui/react-app/public/manifest.json]: Prometheus -> Thanos

File: Makefile [hunk 4]
```diff
@@ -103,34 +109,60 @@ endef
...
+
+.PHONY: react-app-lint-fix
+react-app-lint-fix:
+	@echo ">> running React app linting and fixing errors where possible"
+	cd $(REACT_APP_PATH) && yarn lint
+
+.PHONY: react-app-test
+react-app-test: | $(REACT_APP_NODE_MODULES_PATH) react-app-lint
+	@echo ">> running React app tests"
+	cd $(REACT_APP_PATH) && yarn test --no-watch --coverage
```
Comment (vankop) [inline Makefile]:
```suggestion
	cd $(REACT_APP_PATH) && export CI=true && yarn test --no-watch --coverage
```

Looks like `jest` can not understand that it runs in CI, that's why CI did not pass yet.

**Comment by vankop** [comment]
All unit tests passing.. Don't know how to help with review more..

File: Dockerfile.thanos-ci [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # Available from https://hub.docker.com/r/circleci/golang/
-FROM circleci/golang:1.13.6
+FROM circleci/golang:1.13.6-node
```
Comment (vankop) [inline Dockerfile.thanos-ci]: Did not find which Node.js version runs in this image (Could possibly lead to a build problems later). Better to fix it `package.json` to LTS if possible https://docs.npmjs.com/files/package.json#engines

File: pkg/ui/react-app/src/Navbar.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+              <DropdownItem tag={Link} to={`${pathPrefix}/new/targets`}>
+                Targets
+              </DropdownItem>
+              <DropdownItem tag={Link} to={`${pathPrefix}/new/service-discovery`}>
+                Service Discovery
+              </DropdownItem>
+            </DropdownMenu>
+          </UncontrolledDropdown>
+          <NavItem>
+            <NavLink href="https://prometheus.io/docs/prometheus/latest/getting_started/">Help</NavLink>
```
Comment (vankop) [inline pkg/ui/react-app/src/Navbar.tsx]: Prometheus => Thanos?

**Comment by vankop** [comment]
Could we add typescript check in Makefile and in CI? What do you think @adrien-f ?

package.json
```
scripts: {"types": "tsc --pretty --noEmit"}
```

Right now there are several errors

File: pkg/ui/react-app/src/App.test.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import * as React from 'react';
+import { shallow } from 'enzyme';
+import App from './App';
+import Navigation from './thanos/Navbar';
+import { Container } from 'reactstrap';
+import { Router } from '@reach/router';
+import { Alerts, Config, Flags, Rules, ServiceDiscovery, Status, Targets, TSDBStatus, PanelList } from './pages';
+
+describe('App', () => {
+  const app = shallow(<App pathPrefix="/path/prefix" />);
```
Comment (vankop) [inline pkg/ui/react-app/src/App.test.tsx]: missing `thanosComponent`

**Comment by vankop** [comment]
@bwplotka @adrien-f friendly ping on this 🤗

Are we waiting for more reviewers? or there is other blockers to this PR?

**Comment by GiedriusS** [comment]
> @bwplotka @adrien-f friendly ping on this 
> 
> Are we waiting for more reviewers? or there is other blockers to this PR?

I think this has been postponed for a little bit because in Prometheus this UI code is still very much in flux so we've been recommended to wait for a bit. :-) I guess once it will be moved out of the "experimental" phase we will be able to continue with this. Perhaps @adrien-f knows more information.

**Comment by bwplotka** [comment]
Looking for GSoC help here (: 

cc @prmsrswt

**Comment by GiedriusS** [comment]
Merged via https://github.com/thanos-io/thanos/pull/2412 :P closing this one. Thank you to everyone for your efforts!

**Pull request closed without merging**